core/src/main/java/org/pentaho/di/connections/vfs/provider/ConnectionFileSystem.java

```diff
@@ -31,96 +33,142 @@
+  // When connection does not exist, providerFileObject is null, resulting in a connection file object
+  // for a `!exists()` file.
+  @NonNull
+  protected ConnectionFileObject createFile( @NonNull ConnectionFileName pvfsFileName,
+                                             @Nullable AbstractFileObject<?> providerFileObject ) {
+    return new ConnectionFileObject( pvfsFileName, this, providerFileObject );
+  }
+
+  @Nullable
+  protected AbstractFileObject<?> resolveProviderFile( @NonNull ConnectionFileName pvfsFileName )
+    throws KettleException, IOException {
+    // 1. pvfs:// has no associated connection, so no provider file object.
+    // 2. A connection that does not exist would be rare in this context, but, still, returns null.
+
+    String connectionName = pvfsFileName.getConnectionDecoded();
+
+    VFSConnectionDetails details = vfsConnectionManager.getDetails( connectionName );
+    if ( details == null ) {
+      return null;
     }
-    String url = getUrl( abstractFileName, connectionDetails );
 
-    AbstractFileObject fileObject = null;
-    String domain = null;
+    String providerUri = vfsConnectionManager.getFileNameTransformer( details )
+      .toProviderUri( pvfsFileName, details );
+
+    return (AbstractFileObject<?>) vfsConnectionManager.getKettleVFS()
+      .getFileObject( providerUri, initializeVariableSpace( details ) );
+  }
+
+  @NonNull
+  private <T extends VFSConnectionDetails> VariableSpace initializeVariableSpace( @NonNull T details )
+    throws IOException {
 
-    if ( url != null ) {
-      domain = connectionDetails.getDomain();
-      varSpace.setVariable( CONNECTION, connectionName );
-      fileObject = (AbstractFileObject) KettleVFS.getInstance( bowl ).getFileObject( url, varSpace );
+    VariableSpace varSpace = getSpace();
+
+    // Used by KettleVFSImpl#buildFsOptions -> VFSHelper#getOpts(...).
+    varSpace.setVariable( CONNECTION, details.getName() );
+
+    details.setSpace( varSpace );
+
+    return varSpace;
+  }
+
+  ConnectionFileObject createChild( @NonNull ConnectionFileObject parentFileObject,
+                                    @NonNull FileObject childProviderFileObject )
+    throws FileSystemException {
+
+    @Nullable
+    String connectionName = parentFileObject.getName().getConnectionDecoded();
+
+    ConnectionFileName childPvfsFileName = toPvfsFileName( childProviderFileObject.getName(), connectionName );
+
+    // TODO: check this works, as it is surely much faster... than calling resolveFile( childPvfsUri )
+    // when the provider file object is already available.
+    return createFile( childPvfsFileName, (AbstractFileObject<?>) childProviderFileObject );
+  }
+
+  private ConnectionFileName toPvfsFileName( @NonNull FileName providerFileName, @Nullable String connectionName )
+    throws FileSystemException {
+
+    if ( connectionName == null ) {
+      return new ConnectionFileName( null );
+    }
+
+    VFSConnectionDetails details;
```

Comment:
This method is one of the few places that gets the burden of converting exceptions from `KettleException` to `FileSystemException`.